src/shogun/base/DynArray.h

```diff
@@ -446,16 +446,16 @@ template <class T> class DynArray
 		{
 			for (index_t i=0; i<=current_num_elements-1; ++i)
 				std::swap(
-					array[i], 
+					array[i],
 					array[CMath::random(i, current_num_elements - 1)]);
 		}
 
 		/** randomizes the array with external random state */
-		void shuffle(CRandom * rand)
+		void shuffle(std::shared_ptr<CRandom> rand)
```

Comment:
@vigsterkr these things should be `const std::shared_ptr<T>&` IMO, otherwise you are incrementing the ref count and then decreasing it at the end of the function

Reply:
problem? :)

Reply:
i mean that in this case this should actually work correctly as there's gonna be somebody on the end who holds the ref for this obj.... until that object gets destructed.